bingo/AgeFitness.py

```diff
@@ -0,0 +1,150 @@
+"""Age-Fitness selection
+
+This module implements the Age-Fitness selection algorithm that defines
+the selection used in the Age-Fitness evolutionary algorithm module.
+This module expects to be used in conjunction with the
+``RandomIndividualVariation`` module that wraps the ``VarOr`` module.
+"""
+import numpy as np
+
+from bingo.Base.Selection import Selection
+from bingo.Util.ArgumentValidation import argument_validation
+
+class AgeFitness(Selection):
+    """Age-Fitness selection
+
+    Parameters
+    ----------
+    selection_size : int
+        The size of the group of individuals to be randomly
+        compared. The size must be an integer greater than 1.
+    """
+    WORST_CASE_FACTOR = 50
+
+    @argument_validation(selection_size={">=": 2})
+    def __init__(self, selection_size=2):
+        self._selection_size = selection_size
+        self._selected_indices = []
+        self._population_index_array = np.array([])
+        self._selection_attempts = 0
+
+    @argument_validation(target_population_size={">": 0})
+    def __call__(self, population, target_population_size):
+        """Performs Age-Fitness selection on a population. If ``selection_size``
+        is larger than the population. The maximum number of removed
+        individiuals is the target_population_size.
+
+        Parameters
+        ----------
+        population : list of Chromosome
+            The population on which to perform selection
+        target_population_size : int
+            The size of the new population after selection. It will never be the
+            case that the new population will have a size smaller than the
+            target population.
+
+        Returns
+        -------
+        list of Chromosome :
+            The chromosomes not selected for removal
+
+        Raises
+        ------
+        ValueError
+            If the ``target_population_size`` is larger than the intial
+            `population`
+        """
+        if target_population_size > len(population):
+            raise ValueError("Target population size should\
+                              be less than initial population")
+
+        num_removed = [0]
```

Comment:
`num_removed` is a list to allow the value to be passed by reference rather than value. However, since the only method that writes to `num_removed` is `_remove_individuals(...)`, we can simply return the number that was removed and increment that to a local integer variable.